src/ros_filter.cpp

```diff
@@ -1901,11 +1905,16 @@ namespace RobotLocalization
         clearExpiredHistory(filter_.getLastMeasurementTime() - historyLength_);
       }
 
-      if (!loop_rate.sleep())
+      ros::Duration loop_elapsed_time = ros::Time::now() - loop_end_time;
+      ros::Duration sleep_time = loop_cycle_time - loop_elapsed_time;
+
+      if (!sleep_time.sleep())
```

Comment:
* I've never looked at the ROS code, but I thought the `rate` object attempted to maintain a fixed rate...which seems like what you are trying do. I'm curious why you switched away from it.
* You obviously have a better idea what the `rate`/`sleep` return code means. All the docs say "True if the desired sleep duration was met, false otherwise." I'm guessing a negative duration makes it fail?
* If this check is only there to print an error message, it seems like you would just check if `loop_elapsed_time > loop_cycle_time` (rather than the poorly documented return code from the `sleep` call). You could also use a `ROS_ERROR_COND`, as long you you are breaking out all the strange versions of `ROS_ERROR_XXX`.
* Probably outside the scope, but would this be better/easier inside a `Timer` callback? The Timer will maintain the loop times for you, the measurement callbacks would happen normally and you wouldn't have to worry about `spinOnce` execution time. And the `TimerEvent` gives you some information on when Timer callback was supposed to be called vs when it was actually called.